keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,256 @@
+  * [v1 API](#v1-api)
+* [Test Plan](#test-plan)
+  * [Integration tests for GA](#integration-tests-for-ga)
+  * [e2e tests for GA](#e2e-tests-for-ga)
+  * [Conformance plan for GA](#conformance-plan-for-ga)
+  * [Scale Targets for GA](#scale-targets-for-ga)
+* [Graduation Criteria](#graduation-criteria)
+* [Post-GA tasks](#post-ga-tasks)
+  * [Arbitrary subresources as Post-GA task](#arbitrary-subresources-as-post-ga-task)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+CustomResourceDefinitions (CRDs) are the way to extend the Kubernetes API to
+include custom resource types that behave like the native resource types. CRDs
+have been in Beta since Kubernetes 1.7. This document outlines the required
+steps to graduate CRDs to general availability (GA).
+
+## Motivation
+
+CRDs are in widespread use as a Kubernetes extensibility mechanism and have been
+available in beta for the last 5 Kubernetes releases. Feature requests and bug
+reports suggest CRDs are nearing GA quality, and this KEP aims to itemize the
+remaining improvements to move this toward GA.
+
+### Goals
+
+Establish consensus for the remaining essential improvements needed to move CRDs to GA.
+
+Guiding principles:
+* if a missing feature forces or encourages users to implement non-Kubernetes-like APIs, and therefore damages the ecosystem long term, it belongs on this list.
+* If a feature cannot be added as an after-thought of a GA API, it or some preparation belongs on this list.
+
+
+### Non-Goals
+
+Full feature parity with native kubernetes resource types is not a GA graduation goal. See above guiding principles and below Post-GA tasks.
+
+## Proposal
+
+Objectives to graduate CRDs to GA, each of which is described in more detail further below:
+
+* Defaulting and pruning for custom resources is implemented
+* CRD v1 schemas are restricted to a subset of the OpenAPI specification (and there is a v1beta1 compatibility plan)
+* Generator exists for CRD Validation Schema v3 (Kubebuilder)
+* CustomResourceWebhookConversion API is GA ready
+* CustomResourceSubresources API is GA ready
+
+Bug fixes required to graduate CRDs to GA:
+
+* See “Required for GA” issues tracked via the [CRD Project Board](https://github.com/orgs/kubernetes/projects/28).
+
+For additional details on already completed features, see the [Umbrella Issue](https://github.com/kubernetes/kubernetes/issues/58682).
+
+See [Post-GA tasks](#post-ga-tasks) for decided out-of-scope features.
+
+### Defaulting and pruning for custom resources is implemented
+
+Both defaulting and pruning and also read-only validation a blocked by the
+OpenAPI subset definition (next point). An update of the [old Pruning for
+CustomResources KEP](kubernetes/enhancements#709) and the implementation
+([pruning PR](kubernetes/kubernetes#64558), [defaulting
+PR](kubernetes/kubernetes#63604)), are follow-ups as soon as unblocked.
```

Comment:
the links don't work well, need full URL

Reply:
Oops, fixed.